src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxyLifecycleInterceptor.java

```diff
@@ -51,8 +52,9 @@
    * @param proxy A connection proxy representing a logical connection. This proxy's internal state
    *     may be changed by this class.
    */
-  public ConnectionProxyLifecycleInterceptor(ConnectionProxy proxy) {
+  public ConnectionProxyLifecycleInterceptor(ConnectionProxy proxy, ConnectionPluginManager pluginManager) {
```

Comment:
I think this constructor should accept a single parameter and it should an interface something ILifecycleConsumer that contains two methods transactionBegun and transactionCompleted. Plugin manager, as well as all plugins should implement that interface. Plugin manager is bounded to a current connection so it can synchronize on it.